src/dagmc/DagMC.cpp

```diff
@@ -0,0 +1,774 @@
+  rval = get_group_name(group_set, group_name);
+  if (rval != MB_SUCCESS)
+    return rval;
+
+  std::vector< std::string > group_tokens;
+  tokenize(group_name, group_tokens, delimiters);
+
+  // iterate over all the keyword positions
+  // keywords are even indices, their values (optional) are odd indices
+  for (unsigned int i = 0; i < group_tokens.size(); i += 2) {
+    std::string groupkey = group_tokens[i];
+    std::string groupval;
+    if (i < group_tokens.size() - 1)
+      groupval = group_tokens[i + 1];
+    result[groupkey] = groupval;
+  }
+  return MB_SUCCESS;
+}
+
+ErrorCode DagMC::detect_available_props(std::vector<std::string>& keywords_list, const char* delimiters) {
+  ErrorCode rval;
+  std::set< std::string > keywords;
+  for (std::vector<EntityHandle>::const_iterator grp = group_handles().begin();
+       grp != group_handles().end(); ++grp) {
+    std::map< std::string, std::string > properties;
+    rval = parse_group_name(*grp, properties, delimiters);
+    if (rval == MB_TAG_NOT_FOUND)
+      continue;
+    else if (rval != MB_SUCCESS)
+      return rval;
+
+    for (prop_map::iterator i = properties.begin();
+         i != properties.end(); ++i) {
+      keywords.insert((*i).first);
+    }
+  }
+  keywords_list.assign(keywords.begin(), keywords.end());
+  return MB_SUCCESS;
+}
+
+ErrorCode DagMC::append_packed_string(Tag tag, EntityHandle eh,
+                                      std::string& new_string) {
+  // When properties have multiple values, the values are tagged in a single character array
+  // with the different values separated by null characters
+  ErrorCode rval;
+  const void* p;
+  const char* str;
+  int len;
+  rval = MBI->tag_get_by_ptr(tag, &eh, 1, &p, &len);
+  if (rval == MB_TAG_NOT_FOUND) {
+    // This is the first entry, and can be set directly
+    p = new_string.c_str();
+    return MBI->tag_clear_data(tag, &eh, 1, p, new_string.length() + 1);
+  } else if (rval != MB_SUCCESS)
+    return rval;
+  else {
+    str = static_cast<const char*>(p);
+  }
+
+  // append a new value for the property to the existing property string
+  unsigned int tail_len = new_string.length() + 1;
+  char* new_packed_string = new char[ len + tail_len ];
+  memcpy(new_packed_string, str, len);
+  memcpy(new_packed_string + len, new_string.c_str(), tail_len);
+
+  int new_len = len + tail_len;
+  p = new_packed_string;
+  rval = MBI->tag_set_by_ptr(tag, &eh, 1, &p, &new_len);
+  delete[] new_packed_string;
+  return rval;
+}
+
+ErrorCode DagMC::unpack_packed_string(Tag tag, EntityHandle eh,
+                                      std::vector< std::string >& values) {
+  ErrorCode rval;
+  const void* p;
+  const char* str;
+  int len;
+  rval = MBI->tag_get_by_ptr(tag, &eh, 1, &p, &len);
+  if (rval != MB_SUCCESS)
+    return rval;
+  str = static_cast<const char*>(p);
+  int idx = 0;
+  while (idx < len) {
+    std::string item(str + idx);
+    values.push_back(item);
+    idx += item.length() + 1;
```

Comment:
Readability: By my read, the packed strings are separated by a sentinel of some type which requires a `+ 1` offset? maybe a `NULL`?  If so, I would document it here for readability.

Reply:
@pshriwise @gonuke any idea about this ?
I don't know enough about how the string are packed here to do the right thing...

Reply:
See the comment in line 502 - they are packed in a single string separated by NULL which is how line 544 even works: the new `std::string` is initialized with a `char*` and reads to the NULL character

Reply:
instead of  a comment, you could add a line:

```suggestion
        int null_delimiter_length = 1;
        idx += item.length() + null_delimiter_length;
```

Reply:
I'll add that into my PR #676